pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1631 @@
+		ipvX_addr = "ipv6_addr"
+		noLocalhost = "ip6 daddr != ::1"
+	}
+
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for kube-proxy"),
+	})
+
+	// Create and flush base chains
+	for _, bc := range nftablesBaseChains {
+		chain := &knftables.Chain{
+			Name:     bc.name,
+			Type:     ptr.To(bc.chainType),
+			Hook:     ptr.To(bc.hook),
+			Priority: ptr.To(bc.priority),
+		}
+		tx.Add(chain)
+		tx.Flush(chain)
+	}
+
+	// Create and flush ordinary chains and add rules jumping to them
+	existingChains := sets.New[string]()
+	for _, c := range nftablesJumpChains {
+		ensureChain(c.dstChain, tx, existingChains)
+		tx.Add(&knftables.Rule{
+			Chain: c.srcChain,
+			Rule: knftables.Concat(
+				c.extraArgs,
+				"jump", c.dstChain,
+			),
+		})
+	}
+
+	// Ensure all of our other "top-level" chains exist
+	for _, chain := range []string{kubeServicesChain, kubeForwardChain, kubeMasqueradingChain, kubeMarkMasqChain} {
+		ensureChain(chain, tx, existingChains)
+	}
+
+	// Add the rules in the mark-for-masquerade and masquerading chains
+	tx.Add(&knftables.Rule{
+		Chain: kubeMarkMasqChain,
+		Rule: knftables.Concat(
+			"mark", "set", "mark", "or", proxier.masqueradeMark,
+		),
+	})
+
+	tx.Add(&knftables.Rule{
+		Chain: kubeMasqueradingChain,
+		Rule: knftables.Concat(
+			"mark", "and", proxier.masqueradeMark, "==", "0",
+			"return",
+		),
+	})
+	tx.Add(&knftables.Rule{
+		Chain: kubeMasqueradingChain,
+		Rule: knftables.Concat(
+			"mark", "set", "mark", "xor", proxier.masqueradeMark,
+		),
+	})
+	tx.Add(&knftables.Rule{
+		Chain: kubeMasqueradingChain,
+		Rule:  "masquerade fully-random",
+	})
+
+	// Drop the packets in INVALID state, which would potentially cause
+	// unexpected connection reset if nf_conntrack_tcp_be_liberal is not set.
+	// Ref: https://github.com/kubernetes/kubernetes/issues/74839
+	// Ref: https://github.com/kubernetes/kubernetes/issues/117924
+	if !proxier.conntrackTCPLiberal {
+		tx.Add(&knftables.Rule{
+			Chain: kubeForwardChain,
+			Rule:  "ct state invalid drop",
+		})
+	}
+
+	// Fill in nodeport-ips set if needed (or delete it if not). (We do "add+delete"
+	// rather than just "delete" when we want to ensure the set doesn't exist, because
+	// doing just "delete" would return an error if the set didn't exist.)
+	tx.Add(&knftables.Set{
+		Name:    kubeNodePortIPsSet,
+		Type:    ipvX_addr,
+		Comment: ptr.To("IPs that accept NodePort traffic"),
+	})
+	if proxier.nodePortAddresses.MatchAll() {
+		tx.Delete(&knftables.Set{
+			Name: kubeNodePortIPsSet,
+		})
+	} else {
+		tx.Flush(&knftables.Set{
+			Name: kubeNodePortIPsSet,
+		})
+		nodeIPs, _ := proxier.nodePortAddresses.GetNodeIPs(proxier.networkInterfacer)
+		for _, ip := range nodeIPs {
+			if ip.IsLoopback() {
+				klog.ErrorS(nil, "--nodeport-addresses includes localhost but localhost NodePorts are not supported", "address", ip.String())
+				continue
+			}
+			tx.Add(&knftables.Element{
+				Set: kubeNodePortIPsSet,
+				Key: []string{
+					ip.String(),
+				},
```

Comment:
is it better to add multiple commands or just pass the array of IPs in one command, or it does not really matter?

Reply:
So this is the API being confusing... the `[]string` here doesn't mean you can pass multiple keys; it means you can pass a key that has multiple components.

suggestions welcome...